File: frontend/src/app/core/services/review.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/api-response.model';
import { PageResponse } from '../models/product.model';
import {
  ReviewResponse,
  ProductRatingSummary,
  CreateReviewRequest,
  UpdateReviewRequest
} from '../models/review.model';

@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/reviews`;

  getReviewsByProduct(productId: string, page: number = 0, size: number = 10): Observable<ApiResponse<PageResponse<ReviewResponse>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<ReviewResponse>>>(`${this.apiUrl}/product/${productId}`, { params });
  }

  getProductRatingSummary(productId: string): Observable<ApiResponse<ProductRatingSummary>> {
    return this.http.get<ApiResponse<ProductRatingSummary>>(`${this.apiUrl}/product/${productId}/summary`);
  }

  getMyReviews(page: number = 0, size: number = 10): Observable<ApiResponse<PageResponse<ReviewResponse>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<ReviewResponse>>>(`${this.apiUrl}/my-reviews`, { params });
  }

  createReview(request: CreateReviewRequest): Observable<ApiResponse<ReviewResponse>> {
    return this.http.post<ApiResponse<ReviewResponse>>(this.apiUrl, request);
  }

  updateReview(id: string, request: UpdateReviewRequest): Observable<ApiResponse<ReviewResponse>> {
    return this.http.put<ApiResponse<ReviewResponse>>(`${this.apiUrl}/${id}`, request);
  }

  deleteReview(id: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }

  getAllReviewsAdmin(page: number = 0, size: number = 10): Observable<ApiResponse<PageResponse<ReviewResponse>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<ReviewResponse>>>(`${this.apiUrl}/admin`, { params });
  }
}

File: backend/src/main/java/com/e_commerce/backend/feature_review/service/ReviewService.java
package com.e_commerce.backend.feature_review.service;

import com.e_commerce.backend.feature_review.dto.request.CreateReviewRequest;
import com.e_commerce.backend.feature_review.dto.request.UpdateReviewRequest;
import com.e_commerce.backend.feature_review.dto.response.ProductRatingSummaryResponse;
import com.e_commerce.backend.feature_review.model.ReviewEntity;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface ReviewService {

    ReviewEntity createReview(UUID userId, CreateReviewRequest request);

    Page<ReviewEntity> getReviewsByProduct(UUID productId, Pageable pageable);

    Page<ReviewEntity> getMyReviews(UUID userId, Pageable pageable);

    ReviewEntity getReviewById(UUID id);

    ReviewEntity updateReview(UUID id, UUID userId, UpdateReviewRequest request, boolean isAdmin);

    void deleteReview(UUID id, UUID userId, boolean isAdmin);

    ProductRatingSummaryResponse getProductRatingSummary(UUID productId);
}
